File: apps/mobile/android/app/src/main/java/com/houselevi/plus/data/models/ShopModels.kt
package com.houselevi.plus.data.models

import com.google.gson.annotations.SerializedName

//  Product variant  exact schema 
data class ProductVariant(
    @SerializedName("_id")      val id: String      = "",
    @SerializedName("sku")      val sku: String     = "",
    @SerializedName("title")    val title: String   = "",
    @SerializedName("color")    val color: String?  = null,
    @SerializedName("size")     val size: String?   = null,
    @SerializedName("edition")  val edition: String? = null,
    @SerializedName("price")    val price: Double   = 0.0,
    @SerializedName("stock")    val stock: Int      = 0,
    @SerializedName("barcode")  val barcode: String? = null,
) {
    val inStock: Boolean get() = stock > 0
    val lowStock: Boolean get() = stock in 1..5
    // Label shown in variant chip
    val label: String get() = when {
        size != null && color != null -> "$size / $color"
        size != null    -> size
        color != null   -> color
        edition != null -> edition
        else            -> title
    }
}

//  Product image 
data class ProductImage(
    @SerializedName("url")       val url: String       = "",
    @SerializedName("alt")       val alt: String       = "",
    @SerializedName("isPrimary") val isPrimary: Boolean = false,
    @SerializedName("order")     val order: Int         = 0,
)

//  Product 
data class Product(
    @SerializedName("_id")               val id: String              = "",
    @SerializedName("title")             val title: String           = "",
    @SerializedName("slug")              val slug: String            = "",
    @SerializedName("description")       val description: String     = "",
    @SerializedName("basePrice")         val basePrice: Double       = 0.0,
    @SerializedName("currency")          val currency: String        = "KES",
    @SerializedName("discountPrice")     val discountPrice: Double?  = null,
    @SerializedName("totalStock")        val totalStock: Int         = 0,
    @SerializedName("lowStockThreshold") val lowStockThreshold: Int  = 5,
    @SerializedName("images")           val images: List<ProductImage> = emptyList(),
    @SerializedName("variants")         val variants: List<ProductVariant> = emptyList(),
    @SerializedName("tags")             val tags: List<String>      = emptyList(),
    @SerializedName("isFeatured")       val isFeatured: Boolean     = false,
    @SerializedName("status")           val status: String          = "published",
    @SerializedName("collectionId")     val collectionId: String?   = null,
) {
    val primaryImage: String get() =
        images.firstOrNull { it.isPrimary }?.url
            ?: images.sortedBy { it.order }.firstOrNull()?.url
            ?: ""

    val displayPrice: String get() {
        val p = discountPrice ?: basePrice
        return "$currency ${String.format("%,.0f", p)}"
    }
    val originalPrice: String? get() =
        if (discountPrice != null) "$currency ${String.format("%,.0f", basePrice)}" else null

    val hasDiscount: Boolean get() = discountPrice != null && discountPrice < basePrice
    val inStock: Boolean get() = totalStock > 0
    val isLowStock: Boolean get() = totalStock in 1..lowStockThreshold
}

//  Cart models 
data class CartItem(
    @SerializedName("_id")          val itemId: String     = "",
    @SerializedName("productId")    val productId: String  = "",
    @SerializedName("variantId")    val variantId: String  = "",
    @SerializedName("variantTitle") val variantTitle: String = "",
    @SerializedName("title")        val title: String      = "",
    @SerializedName("price")        val price: Double      = 0.0,
    @SerializedName("currency")     val currency: String   = "KES",
    @SerializedName("quantity")     val quantity: Int      = 1,
    @SerializedName("imageUrl")     val imageUrl: String   = "",
) {
    val lineTotal: Double get() = price * quantity
    val displayPrice: String get() = "$currency ${String.format("%,.0f", price)}"
    val displayTotal: String get() = "$currency ${String.format("%,.0f", lineTotal)}"
}

data class Cart(
    @SerializedName("items")    val items: List<CartItem> = emptyList(),
    @SerializedName("subtotal") val subtotal: Double      = 0.0,
    @SerializedName("currency") val currency: String      = "KES",
) {
    val isEmpty: Boolean get() = items.isEmpty()
    val itemCount: Int get() = items.sumOf { it.quantity }
    val displaySubtotal: String get() = "$currency ${String.format("%,.0f", subtotal)}"
}

//  Request bodies 
data class AddToCartBody(
    val productId: String,
    val variantId: String,
    val variantTitle: String,
    val price: Double,
    val quantity: Int = 1,
)

data class UpdateQuantityBody(val quantity: Int)

data class CreateOrderBody(
    val phone: String,
    val shippingAddress: ShippingAddress,
    val shippingLocation: String,
    val currency: String = "KES",
)

data class ShippingAddress(
    val line1: String,
    val city: String,
    val country: String,
    val postalCode: String = "",
)

//  API responses 
data class ProductsResponse(
    @SerializedName("success") val success: Boolean        = false,
    @SerializedName("data")    val data: List<Product>     = emptyList(),
)
data class ProductResponse(
    @SerializedName("success") val success: Boolean = false,
    @SerializedName("data")    val data: Product?   = null,
)
data class CartResponse(
    @SerializedName("success") val success: Boolean = false,
    @SerializedName("data")    val data: Cart?      = null,
    // Some endpoints return cart directly
    @SerializedName("items")    val items: List<CartItem>? = null,
    @SerializedName("subtotal") val subtotal: Double?      = null,
) {
    val cart: Cart get() = data ?: Cart(
        items    = items    ?: emptyList(),
        subtotal = subtotal ?: 0.0,
    )
}
data class OrderResponse(
    @SerializedName("success")    val success: Boolean = false,
    @SerializedName("orderId")    val orderId: String  = "",
    @SerializedName("pesapalUrl") val pesapalUrl: String = "",
)

File: apps/mobile/ios/HouseLevi+/Models/ShopModels.swift
import Foundation

struct ProductVariant: Codable, Identifiable, Equatable {
    var id: String = ""
    var sku: String = ""; var title: String = ""
    var color: String?; var size: String?; var edition: String?
    var price: Double = 0; var stock: Int = 0; var barcode: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"; case sku; case title; case color; case size
        case edition; case price; case stock; case barcode
    }
    init() {}
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id      = (try? c.decodeIfPresent(String.self, forKey: .id))      ?? ""
        sku     = (try? c.decodeIfPresent(String.self, forKey: .sku))     ?? ""
        title   = (try? c.decodeIfPresent(String.self, forKey: .title))   ?? ""
        color   = try? c.decodeIfPresent(String.self,  forKey: .color)
        size    = try? c.decodeIfPresent(String.self,  forKey: .size)
        edition = try? c.decodeIfPresent(String.self,  forKey: .edition)
        price   = (try? c.decodeIfPresent(Double.self, forKey: .price))   ?? 0
        stock   = (try? c.decodeIfPresent(Int.self,    forKey: .stock))   ?? 0
        barcode = try? c.decodeIfPresent(String.self,  forKey: .barcode)
    }
    var inStock: Bool { stock > 0 }
    var lowStock: Bool { stock > 0 && stock <= 5 }
    var label: String {
        if let s = size, let c = color { return "\(s) / \(c)" }
        if let s = size    { return s }
        if let c = color   { return c }
        if let e = edition { return e }
        return title
    }
}

struct ProductImage: Codable, Equatable {
    var url: String = ""; var alt: String = ""
    var isPrimary: Bool = false; var order: Int = 0
}

struct Product: Codable, Identifiable, Equatable {
    var id: String = ""; var title: String = ""; var slug: String = ""
    var description: String = ""
    var basePrice: Double = 0; var currency: String = "KES"
    var discountPrice: Double?; var totalStock: Int = 0
    var lowStockThreshold: Int = 5
    var images: [ProductImage] = []; var variants: [ProductVariant] = []
    var tags: [String] = []; var isFeatured: Bool = false; var status: String = "published"

    enum CodingKeys: String, CodingKey {
        case id = "_id"; case title; case slug; case description
        case basePrice; case currency; case discountPrice; case totalStock
        case lowStockThreshold; case images; case variants; case tags; case isFeatured; case status
    }
    init() {}
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id               = (try? c.decodeIfPresent(String.self,           forKey: .id))               ?? ""
        title            = (try? c.decodeIfPresent(String.self,           forKey: .title))            ?? ""
        slug             = (try? c.decodeIfPresent(String.self,           forKey: .slug))             ?? ""
        description      = (try? c.decodeIfPresent(String.self,           forKey: .description))      ?? ""
        basePrice        = (try? c.decodeIfPresent(Double.self,           forKey: .basePrice))        ?? 0
        currency         = (try? c.decodeIfPresent(String.self,           forKey: .currency))         ?? "KES"
        discountPrice    = try? c.decodeIfPresent(Double.self,            forKey: .discountPrice)
        totalStock       = (try? c.decodeIfPresent(Int.self,              forKey: .totalStock))       ?? 0
        lowStockThreshold = (try? c.decodeIfPresent(Int.self,             forKey: .lowStockThreshold)) ?? 5
        images           = (try? c.decodeIfPresent([ProductImage].self,   forKey: .images))           ?? []
        variants         = (try? c.decodeIfPresent([ProductVariant].self, forKey: .variants))         ?? []
        tags             = (try? c.decodeIfPresent([String].self,         forKey: .tags))             ?? []
        isFeatured       = (try? c.decodeIfPresent(Bool.self,             forKey: .isFeatured))       ?? false
        status           = (try? c.decodeIfPresent(String.self,           forKey: .status))           ?? "published"
    }

    var primaryImage: String {
        images.first(where: { $0.isPrimary })?.url
            ?? images.sorted(by: { $0.order < $1.order }).first?.url ?? ""
    }
    var displayPrice: String {
        let p = discountPrice ?? basePrice
        return "\(currency) \(String(format: "%.0f", p))"
    }
    var originalPrice: String? {
        guard let d = discountPrice, d < basePrice else { return nil }
        return "\(currency) \(String(format: "%.0f", basePrice))"
    }
    var hasDiscount: Bool { discountPrice != nil && (discountPrice ?? 0) < basePrice }
    var inStock: Bool { totalStock > 0 }
    var isLowStock: Bool { totalStock > 0 && totalStock <= lowStockThreshold }
}

struct CartItem: Codable, Identifiable, Equatable {
    var id: String = ""
    var productId: String = ""; var variantId: String = ""; var variantTitle: String = ""
    var title: String = ""; var price: Double = 0; var currency: String = "KES"
    var quantity: Int = 1; var imageUrl: String = ""
    enum CodingKeys: String, CodingKey {
        case id = "_id"; case productId; case variantId; case variantTitle
        case title; case price; case currency; case quantity; case imageUrl
    }
    init() {}
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id           = (try? c.decodeIfPresent(String.self, forKey: .id))           ?? ""
        productId    = (try? c.decodeIfPresent(String.self, forKey: .productId))    ?? ""
        variantId    = (try? c.decodeIfPresent(String.self, forKey: .variantId))    ?? ""
        variantTitle = (try? c.decodeIfPresent(String.self, forKey: .variantTitle)) ?? ""
        title        = (try? c.decodeIfPresent(String.self, forKey: .title))        ?? ""
        price        = (try? c.decodeIfPresent(Double.self, forKey: .price))        ?? 0
        currency     = (try? c.decodeIfPresent(String.self, forKey: .currency))     ?? "KES"
        quantity     = (try? c.decodeIfPresent(Int.self,    forKey: .quantity))     ?? 1
        imageUrl     = (try? c.decodeIfPresent(String.self, forKey: .imageUrl))     ?? ""
    }
    var lineTotal: Double { price * Double(quantity) }
    var displayPrice: String { "\(currency) \(String(format: "%.0f", price))" }
    var displayTotal: String { "\(currency) \(String(format: "%.0f", lineTotal))" }
}

struct Cart: Codable, Equatable {
    var items: [CartItem] = []; var subtotal: Double = 0; var currency: String = "KES"
    var isEmpty: Bool { items.isEmpty }
    var itemCount: Int { items.reduce(0) { $0 + $1.quantity } }
    var displaySubtotal: String { "\(currency) \(String(format: "%.0f", subtotal))" }
}

struct ProductsResponse: Codable { var success: Bool = false; var data: [Product] = [] }
struct ProductResponse: Codable  { var success: Bool = false; var data: Product? }
struct CartResponse: Codable {
    var success: Bool = false; var data: Cart?
    var items: [CartItem]?; var subtotal: Double?
    var cart: Cart { data ?? Cart(items: items ?? [], subtotal: subtotal ?? 0) }
}
struct OrderResponse: Codable {
    var success: Bool = false; var orderId: String = ""; var pesapalUrl: String = ""
}
struct AddToCartBody: Encodable {
    var productId: String; var variantId: String; var variantTitle: String
    var price: Double; var quantity: Int
}
struct UpdateQuantityBody: Encodable { var quantity: Int }
struct CreateOrderBody: Encodable {
    var phone: String
    var shippingAddress: ShippingAddress
    var shippingLocation: String
    var currency: String = "KES"
}
struct ShippingAddress: Encodable { var line1: String; var city: String; var country: String; var postalCode: String = "" }
